compiler/rustc_session/src/config.rs

```diff
@@ -1656,7 +1658,7 @@ pub fn parse_externs(
     for arg in matches.opt_strs("extern") {
         let (name, path) = match arg.split_once('=') {
             None => (arg, None),
-            Some((name, path)) => (name.to_string(), Some(path.to_string())),
+            Some((name, path)) => (name.to_string(), Some(PathBuf::from(path))),
```

Comment:
I think `Path::new` would also work. At least if you assign the result of `arg.split_once('=')` to a variable. This would avoid a single allocation.

Reply:
Not sure I understand what you mean. `path` is a `&str` so we need to allocate once. I think `PathBuf::from` is the best way to achieve this.

Reply:
I think they meant something along the lines of:

```diff
diff --git a/compiler/rustc_session/src/config.rs b/compiler/rustc_session/src/config.rs
index 3352f5fbce9..529ea6bb47b 100644
--- a/compiler/rustc_session/src/config.rs
+++ b/compiler/rustc_session/src/config.rs
@@ -1656,7 +1656,7 @@ pub fn parse_externs(
     for arg in matches.opt_strs("extern") {
         let (name, path) = match arg.split_once('=') {
             None => (arg, None),
-            Some((name, path)) => (name.to_string(), Some(PathBuf::from(path))),
+            Some((name, path)) => (name.to_string(), Some(Path::new(path))),
         };
         let (options, name) = match name.split_once(':') {
             None => (None, name),
@@ -1671,7 +1671,7 @@ pub fn parse_externs(
             // --extern prelude_name=some_file.rlib
             match entry {
                 Entry::Vacant(vacant) => {
-                    let path = std::fs::canonicalize(&path).unwrap_or_else(|_| path);
+                    let path = std::fs::canonicalize(&path).unwrap_or_else(|_| path.to_owned());
                     let files = BTreeSet::from_iter(iter::once(path));
                     vacant.insert(ExternEntry::new(ExternLocation::ExactPaths(files)))
                 }
@@ -1679,14 +1679,14 @@ pub fn parse_externs(
                     let ext_ent = occupied.into_mut();
                     match ext_ent {
                         ExternEntry { location: ExternLocation::ExactPaths(files), .. } => {
-                            files.insert(path);
+                            files.insert(path.to_owned());
                         }
                         ExternEntry {
                             location: location @ ExternLocation::FoundInLibrarySearchDirectories,
                             ..
                         } => {
                             // Exact paths take precedence over search directories.
-                            let path = std::fs::canonicalize(&path).unwrap_or_else(|_| path);
+                            let path = std::fs::canonicalize(&path).unwrap_or_else(|_| path.to_owned());
                             let files = BTreeSet::from_iter(iter::once(path));
                             *location = ExternLocation::ExactPaths(files);
                         }
```

which works out to be _some_ improvement given the fact that `std::fs::canonicalize` will allocate for you later in some of the branches.

Reply:
Ah yes! This should save an allocation on a vast majority runs of this code as I suspect that most of these paths will be successfully canonicalized which does allocate.